### PR TITLE
fix: extract write strategy

### LINKED ISSUE #36: Extract a shared Open Table Format write strategy in loader
## Problem statement
Iceberg support was added with parity-oriented branching that mirrors Delta behavior in `src/loader/s3_loader.py`, but table-format logic is still spread across multiple format-specific methods (`_load_delta`, `_load_iceberg`, merge helpers, existence checks, and path/identifier conversion). This increases duplication and makes future table formats (or behavior changes) harder to add safely.

## Why this matters
- More branching increases regression risk when changing write modes (`append`, `overwrite`, `merge`).
- Shared semantics (table path resolution, table existence, dedup, schema handling) can drift between Delta and Iceberg.
- Extensibility for future Open Table Formats is constrained by ad-hoc conditional logic.

## Scope
**In scope:**
- Introduce an explicit table-format strategy boundary (e.g., protocol/base class + Delta/Iceberg implementations) in loader code.
- Centralize common table write flow in one orchestration path:
  - input normalization
  - duplicate column handling / sanitization
  - optional merge dedup behavior
  - mode routing (`append`/`overwrite`/`merge`)
- Move format-specific operations behind strategy methods (e.g., table identifier resolution, merge execution, metadata existence check).
- Keep existing external config contract unchanged.

**Out of scope:**
- Supporting additional formats beyond Delta/Iceberg in this issue.
- Behavioral changes to merge semantics beyond code-structure refactor.

## Proposed approach
- Add a table-format abstraction under `src/loader/` used by `S3Loader`.
- Convert current helpers (`_generate_table_path`, `_table_exists`, `_perform_delta_merge`, `_perform_iceberg_merge`) into shared-vs-strategy responsibilities.
- Make `S3Loader.load` delegate table-format behavior through a single centralized path.
- Add focused tests for strategy parity (same mode behavior contracts where feasible).

## Done when
- [x] `src/loader/s3_loader.py` no longer contains duplicated mode orchestration for Delta vs Iceberg.
- [x] Delta and Iceberg-specific logic lives behind strategy-level APIs.
- [x] Existing Delta behavior remains compatible.
- [x] Iceberg behavior from issue #15 remains compatible.
- [x] Tests cover centralized flow and strategy-specific paths.

## Dependencies / sequencing
- Depends on completion of #15 (Iceberg write support parity).
- Should ship before adding any third table format.

### LINKED ISSUE #37: Normalize table identity and existence semantics (path vs catalog)
## Problem statement
Current Iceberg behavior derives catalog table names from filesystem paths and detects table existence via filesystem metadata (`metadata` directory). Delta and Iceberg now share some helpers, but their table identity model remains implicit and inconsistent (filesystem-native vs catalog-native semantics).

## Why this matters
- Mixed identity models are fragile (especially with non-trivial warehouse roots, naming edge cases, and quoting).
- Existence checks based only on filesystem markers can diverge from catalog state.
- Future integrations (alternate catalogs, external table management) need a clear abstraction for “table identity”.

## Scope
**In scope:**
- Define explicit table identity model per format (e.g., path identifier + optional catalog identifier).
- Centralize conversion helpers and validation for warehouse-relative identifiers.
- Introduce format-aware existence checks that prefer authoritative metadata source for each format.
- Document invariants for `prefix`, warehouse root, and identifier derivation.

**Out of scope:**
- Adding non-Hadoop Iceberg catalog support (REST/Glue/Hive) in this issue.
- Broad rename/migration of persisted table locations.

## Proposed approach
- Add a small `TableIdentity` value object (or equivalent) that captures resolved location and logical identifier.
- Refactor Iceberg identifier derivation into dedicated tested helper module.
- Make existence checking strategy-based and format-owned.
- Add edge-case tests for unusual prefixes and warehouse roots.

## Done when
- [x] Path→identifier derivation no longer lives as ad-hoc logic in loader write methods.
- [x] Table existence checks are explicit per format and tested.
- [x] Failure messages for invalid identity derivation are actionable.
- [x] Docs explain identity semantics without duplicating implementation details.

## Dependencies / sequencing
- Can be parallel with #36  , but cleaner after #36  strategy extraction lands.
- Supports future catalog expansion work.

### PR DESCRIPTION
**Pull request by naineel1209**
Resolves #36 

## Summary

This PR refactors table-format loading so Delta and Iceberg behavior is owned by their respective `LoadStrategy` implementations instead of being handled inside `ObjectStoreLoader`.

Key changes:

- Moved parity-oriented Delta/Iceberg-specific logic out of `ObjectStoreLoader` and into concrete load strategies.
- Consolidated shared table-format write orchestration in `BaseLoadStrategy`, including:
  - write-mode validation
  - append / overwrite / merge routing
  - merge-key validation
  - missing-target bootstrap behavior for merge writes
  - table-location resolution
- Introduced clearer ownership boundaries between shared loading flow and format-specific behavior.
- Abstracted format-owned operations behind strategy hooks, including:
  - table existence checks
  - simple writes
  - merge execution
  - catalog/path-specific target resolution
- Added path-backed Delta loading support where the table location is the canonical write and merge target.
- Added catalog-backed Iceberg loading support where writes and merges operate through Spark catalog identifiers while preserving object-store table locations.
- Updated documentation for the load strategy architecture, naming conventions, and expected ownership boundaries.
- Added tests covering the refactored strategy behavior and format-specific write/existence flows.
- Added or updated test coverage for:
  - `ObjectStoreLoader` 

## Design Notes

Checkout `docs/load_strategy/README.md`

## Checklist

- [x] I have described the change clearly enough for reviewers.
- [x] I ran lint and tests locally (or noted why not).
- [x] This PR does not include secrets, credentials, or internal-only endpoints in YAML, SQL, or docs.

## Related

- Might resolve #37 as well

### CODE REVIEW

File: src/loader/object_store_loader.py [hunk 1]
```diff
@@ -539,65 +343,29 @@ def load(
...
-        metadata_directories = {
-            LoadingFormat.DELTA: "_delta_log",
-            LoadingFormat.ICEBERG: "metadata",
-        }
-        metadata_dir = metadata_directories.get(table_format)
-
-        if not metadata_dir:
-            return False
-
-        try:
```
Comment (victorlou) [inline src/loader/object_store_loader.py:586]: We were checking for a metadata directory under the table location before and now Tte refactor uses spark.catalog.tableExists(...) after setting the warehouse. Is that the same or am i confusing it? e.g. files on disk without a catalog entry, or the opposite. If anything relies on filesystem-only detection (e.g. backfill / destination_exists), I just wonder if this means it will execute differently.

Let me know, this might still be OK, just want to make sure

File: src/load_strategy/delta_strategy.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Union
+
+try:
+    import delta.tables as delta_tables
+except ImportError:  # pragma: no cover - exercised only when optional dependency is absent
+    DeltaTable: Any = None
+else:
+    DeltaTable: Any = delta_tables.DeltaTable
```
Comment (victorlou) [inline src/load_strategy/delta_strategy.py]: Do we need this try except on the delta module? Do think we'd better keep “fail at import” semantics, or validate once when format is Delta? Again, you can stand by this if you think it's right, I was get confused with try excepts on imports to be honest 😅

File: src/utils/transient_storage_retry.py
Comment (victorlou) [inline src/utils/transient_storage_retry.py:1]: Bit confused on the name of the file here, I think it's already implemented as a generic decorator, the "s3_" is misleading.

Also quick sidenote that I keep thinking we're growing and growing the "utils" section of the repo. Nothing to do about it now necessarily, but I wonder how we should deal with it if it ever becomes a gigantic list of things. Might create an issue later for us to review it all

File: docs/load_strategy/README.md
Comment (victorlou) [inline docs/load_strategy/README.md:1]: Worth reviewing if we need to update any other documentation. This doc is solid, but i think we still need to update:
- docs/configuration/loading.md
- docs/configuration/overview.md (just maybe, depending on what we do on the first one)

**Review by victorlou — CHANGES_REQUESTED** [review]

File: src/loader/object_store_loader.py (hunk 1 above)
Comment (naineel1209) [reply src/loader/object_store_loader.py:586]: For `IcebergStrategy`, since everywhere we were using table identifiers like `iceberg.<path_to_table>`, checking if the table exists at `<path_to_table>/metadata` felt a bit odd. Everything else works the same as before, its just that we need to set and unset the warehouse location for the `iceberg` catalog at runtime.

File: docs/load_strategy/README.md
Comment (naineel1209) [reply docs/load_strategy/README.md:1]: added additional documentation in 209757e2d101f877106dd8b66db6794bc82d4771

File: src/utils/transient_storage_retry.py
Comment (naineel1209) [reply src/utils/transient_storage_retry.py:1]: fixed in [53e21fe](https://github.com/victorlou/spine/pull/46/commits/53e21fe4f42c29042d7db9387c8491a7e09ba899)

File: src/load_strategy/delta_strategy.py (hunk 2 above)
Comment (naineel1209) [reply src/load_strategy/delta_strategy.py]: moved to lazy import in [53e21fe](https://github.com/victorlou/spine/pull/46/commits/53e21fe4f42c29042d7db9387c8491a7e09ba899)

**Review by victorlou — APPROVED** [review]
